src/drivers/gitlab.js

```diff
@@ -59,9 +59,7 @@ class Gitlab {
         })
     );
 
-    this.detectedBase = possibleBases.find(
-      (base) => base.constructor !== Error
-    );
+    this.detectedBase = possibleBases.find((base) => typeof base === 'string');
```

Comment:
Could we use?

```js
filter(base => base.startsWith('http'))
```

Reply:
Abracadabra? https://github.com/iterative/cml/issues/596#issuecomment-862491282 🧙🏼‍♂️ 

```js
filter(base => base && base.startsWith && base.startsWith('http'))
```

After #597 it could even be written like this:

```js
filter(base => base?.startsWith?.('http'))
```

Reply:
even better

```js
filter(base => typeof base === 'string'  && base.startsWith('http'))
```

Reply:
> _Could we use?_
> 
> ```js
> filter(base => base && base.startsWith && base.startsWith('http'))
> ```

Assuming that you mean `find` and not `filter` as the used method, yes. Your suggestion seems to cover every case I can think of, even if the intent is slightly different.

Reply:
```js
find(base => typeof base === 'string'  && base.startsWith('http'))
```

Reply:
If we check **begins with `http`** instead of **is not an error**, we would be introducing an additional constraint. Even if it seems apparently harmless and conveys a more explicit intention, I don't know if we might be hitting edge cases where it would fail, like `Https` (note the uppercase `H` character).

Reply:
```js
find(base => typeof base === 'string'  && base.toLowerCase().startsWith('http'))
```

Reply:
We only return a string after receiving a well-formed API response, and we shouldn't need to check again on our own whether the protocol is HTTP or not.

https://github.com/iterative/cml/blob/c9b006038c34023eef837c4d995223c0c22a6dc8/src/drivers/gitlab.js#L50-L58

Reply:
Anyhow, if you consider that there is a valid reason to add this additional check, I'm open to it.

Reply:
nope, its totally wrong. they are pathnames, they dont even start with any protocol.

Reply:
I do not fully understand how you got ```ttp``` there

Reply:
I still think that the error is something related to http vs https maybe?

if they give you a repo like 'http://something' we are going to hit that 'ttp' because of the assumed ```https``` in CML.js and the ```replace().substr(1)```

Reply:
> _nope, it's totally wrong. they are pathnames, they don't even start with any protocol._

Each string on the `possibleBases` array is a full URL that's proven to be a GitLab API base, and all of them do start with a protocol like `http` or `https` as you can see by debugging the code.

https://github.com/iterative/cml/blob/c9b006038c34023eef837c4d995223c0c22a6dc8/src/drivers/gitlab.js#L43

> _I do not fully understand how you got `ttp` there_
> _if they give you a repo like 'http://something' we are going to hit that 'ttp' because of the assumed `https` in CML.js and the `replace().substr(1)` ?_

Before this pull request, the method to get the first valid API base address from `possibleBases` was the following:

https://github.com/iterative/cml/blob/6c451d965aaea2ad2b1812dd2a223def9ce87739/src/drivers/gitlab.js#L62-L64

Unfortunately, when `base` is a `FetchError` object, the `base.constructor !== Error` check doesn't have the desired effect and `this.detectedBase` will get assigned to the first `FetchError` on the `possibleBases` array.

https://github.com/iterative/cml/blob/6c451d965aaea2ad2b1812dd2a223def9ce87739/src/drivers/gitlab.js#L30-L37

Trying to replace `FetchError: a long description of the error` by an empty string on the code above would leave the original `this.repo` string intact. Then, you remove the first character of the result with `.substr(1)` so `this.repo` loses its first character.

Reply:
> _I still think that the error is something related to http vs https maybe?_

Users may have any connection error, and all of them will be masked under an unhelpful `TypeError: Only absolute URLs are supported` until we merge this pull request. HTTP versus HTTPS errors fall in this category.

Reply:
> TypeError: Only absolute URLs are supported

Is this PR improving a better message 😃

Reply:
> _Is this PR improving a better message_

It doesn't improve any message _per se,_ it just avoids masking a helpful error behind an unhelpful one.